### ISSUE TITLE
Different results from compiling less with Grunt or with "server side compilation"

### ISSUE THREAD

**Issue by slackerzz** [magento/magento2#7231]
### Preconditions (*)
1. Magento 2.3.x (tried both CE and EE) 
2. Ubuntu 16.04, PHP 7.x, node v4.6.1

### Steps to reproduce (*)
1. set Magento in developer mode
2. in the backend set css generation in "server side compilation"

While developing a custom template i found some differences between my local dev site (inside a vagrant box) and our staging server. Some css rules in the staging server are not applied, while in my dev box they work as expected. In my local environment i use grunt while in our staging server we use server side compilation.
The produced css files are different, the order in which css rules are processed is different!

So today i made some tests with a clean M2 installations and Luma theme.

| Server Side compilation | Grunt | Diff |
| --- | --- | --- |
| [styles-m.css](https://gist.github.com/slackerzz/a01a156a6ef9c18d5f83f0eb7a3fcb78) | [styles-m.css](https://gist.github.com/slackerzz/70bfd22298caff9e2f9de1793a908c95) | [diff](https://www.diffchecker.com/MmH1jofC)* |
| [styles-l.css](https://gist.github.com/slackerzz/7b70c057817a97afa25c01815e654ba2) | [styles-l.css](https://gist.github.com/slackerzz/b9d44a41cd4562478046a1cce62ab2d1) | [diff](https://www.diffchecker.com/E5pxAVfW)* |

*on the left side of the diff you have the server side compiled css

### Expected result (*)
No diff
### Actual result (*)

As you can see `styles-m.css` is ok, there are no breaking differences between the two versions.
Sadly the two versions of `styles-l.css` are not equivalent, the order of the css rules in the two files is different and this can produce a different result.

**Comment by hostep** [comment]
If you are able to install and use nodejs on your production server, you can use [our module](https://github.com/baldwin-agency/magento2-module-less-js-compiler), which uses less.js to compile less files on the server instead of the less.php library (which indeed produces slightly different results).

**Comment by slackerzz** [comment]
thank you @hostep, i'll try your module

**Comment by slackerzz** [comment]
@hostep your module works well
However this is a major flaw in the official frontend dev workflow

**Comment by tkacheva** [comment]
@slackerzz  Thanks for reporting that. Internal issue MAGETWO-60689

**Comment by slackerzz** [comment]
any news here?

**Comment by orlangur** [comment]
This is expectable by itself since client-side and server-side preprocessing use different library versions.

Providing a reproducible narrowed use case (some visual bug) would simplify investigation.

**Comment by magento-engcom-team** [comment]
@slackerzz, thank you for your report.
We've created internal ticket(s) MAGETWO-60689 to track progress on the issue.

**Comment by joshdavenport** [comment]
Am I living in a bubble or is nobody else other than those already partaking in this issue experiencing this?

I have what seems to be a reproducable case, and am happy to pass on some details if anybody is open to working on this.

**Comment by orlangur** [comment]
Hi @joshdavenport, of course providing a simple test case will simplify the investigation to whoever will work on this.

Note that

> This is expectable by itself since client-side and server-side preprocessing use different library versions.

Magento is not doing LESS compilation by itself, in case of bug most likely we will have to patch `lessphp` library.

**Comment by joshdavenport** [comment]
Before proceeding here, I've actually discovered that the different results for me between LESS complied via grunt and "server side compilation" was eliminated by downgrading grunt-contrib-less from 1.4.1 (which is using less 2.7.1) to 0.12.0 (which is using less 1.7.2). I think @hostep may have discovered this (based on the comments in [their module](https://github.com/baldwin-agency/magento2-module-less-js-compiler#remarks) referenced earlier). grunt-contrib-less is specified as 1.4.1 in the provided package.json.sample so it's fair to assume others using the grunt workflow have a real possibility of stumbling into this issue.

Is it worth me reporting this separately?

**Comment by hostep** [comment]
@joshdavenport : please do so. This happend when upgrading from Magento 2.1 to 2.2, the nodejs dependencies got updated and now less 2.x is getting installed even though the PHP library to compile less is only compatible with less 1.x

I've reported this through some other channel a long time ago, can't remember when or how, but I failed at reporting it here on github. So please, go ahead and open a new issue.

**Comment by slackerzz** [comment]
I opened this issue while i was working with Magento 2.1.2 so it was present before any upgrade to 2.2.
The problem, i think, is the css splitting between `style-m.css` and `style-l.css`. 

In the themes i develop i do not use the `.media_width()` mixin and set them to use and generate a single css file, that's my workaround.

**Comment by mverstraete** [comment]
I may have tracked this down to an issue in grunt-contrib-less v1.4.1, where it is not compiling nested extends.

I had this problem specifically with the Add To Wish List button.  In the Blank theme, the wishlist button gets its button styles via:
`&:extend(.abs-action-addto-product all);`

.abs-action-addto-product is defined in _extends.less, and it contains another extend:  
`&:extend(.abs-action-link-button all);`

.abs-action-link-button, which contains the button styles, is never added to the wishlist button when compiling with grunt-contrib-less v1.4.1.   

it _does_ work if I downgrade to grunt-contrib-less v0.12.0, and it also _does_ work when LESS is compiled with PHP.

So, it seems there is an issue with grunt-contrib-less v1.4.1 and compiling extends within extends.

**Comment by orlangur** [comment]
> different results for me between LESS complied via grunt and "server side compilation" was eliminated by downgrading grunt-contrib-less from 1.4.1 (which is using less 2.7.1) to 0.12.0 (which is using less 1.7.2)

> when upgrading from Magento 2.1 to 2.2, the nodejs dependencies got updated and now less 2.x is getting installed even though the PHP library to compile less is only compatible with less 1.x

@joshdavenport @hostep @mveeramneni can somebody contribute such downgrade of JS library maybe?

**Comment by hostep** [comment]
> even though the PHP library to compile less is only compatible with less 1.x

I'm no longer convinced of this statement I made earlier. If we have to believe the less version specified in the `oyejorge/less.php` library, it is marked as compatible with less 2.5.3 in the latest version: https://github.com/oyejorge/less.php/blob/v1.7.0.14/lib/Less/Version.php#L12

This was increased from less version 1.7.5 to 2.5.3 in version v1.7.0.12 of `oyejorge/less.php` by this commit: https://github.com/oyejorge/less.php/commit/712473a175039a121a535e621b24d03dc8143cf0#diff-c2bd6b044fb355e9bc8f661b46eff114

v1.7.0.12 [was released](https://github.com/oyejorge/less.php/releases/tag/v1.7.0.12) on 6 February 2017 which was in-between Magento 2.2.2 and 2.2.3, so Magento 2.2.0 initially shipped with a nodejs package compatible with less v2 which was probably wrong at that time, but in the mean time it is probably fine by upgrading to the latest releases of the `oyejorge/less.php` package.

I have to little deep knowledge about the differences between v1 and v2 of the less spec to verify if anything of this is true or not. If anybody with more knowledge about this could weigh in, that would be appreciated :)

@mverstraete: what version of the `oyejorge/less.php` package are you using?

**Comment by Brengineer** [comment]
I've noticed a similar issue, with the primary difference being that `.abs-` styles which extend other `.abs-` styles in _extends.less are not output properly when compiled with Grunt, due to being nested (e.g. `.abs-action-addto-product`, which extends `.abs-action-link-button`). This creates differences between production and development sites, which make some style issues invisible in a standard development workflow.

**Comment by ghost** [comment]
It is not an issue the difference between rendered css is:
 - grunt uses shorthand colors where possible instead of full-length hexa-colors. There is css-comb library in npm that can solve this, because grunt-contrib-less doesn't provide such options, look here https://github.com/csscomb/csscomb.js/blob/dev/doc/configuration.md

- grunt doesn't prettifies the output so we rules order is defined by natural files precedence, while server-side library is set to put tag rules before class rules. Solution for this also should exist.

The difference in the example doesn't make any impact on page rendering.

**Comment by joshdavenport** [comment]
That statement seems to directly contradict what everybody has found in this issue, especially https://github.com/magento/magento2/issues/7231#issuecomment-431164523

I personally originally brought this up when I saw legitimately different appearances (due to missing styles) in the cart

**Comment by mverstraete** [comment]
@joshdavenport is correct.  The issue is 100% the fact that grunt-contrib-less v1.4.1 does not compile nested extends, while the php less compiler does.

**Comment by slackerzz** [comment]
Looking at the [diff](https://www.diffchecker.com/E5pxAVfW#right-7) that i linked 2 years ago you can see that both **php and grunt compile the same set of rules**, but the order is different.

In the diff the css in the left comes from php, the right one from grunt.
Take this random css rule, that comes from https://github.com/magento/magento2/blob/2.3-develop/app/design/frontend/Magento/luma/web/css/source/_extends.less#L478:
```css
.abs-margin-for-forms-desktop {
     margin-left: 25.8%;
}
```

The php compiler put this rule at line 101 while grunt at 561.

Another example that comes from from https://github.com/magento/magento2/blob/2.3-develop/app/design/frontend/Magento/luma/web/css/source/_extends.less#L159:
```css
.abs-button-desktop {		
    width: auto;		
}
```
The php compiler put this rule at line 27 while grunt at 487.

**Comment by mverstraete** [comment]
@slackerzz , indeed, the solution to your original problem is simply that the order of css rules is different between PHP and Grunt.  

There is a second issue being talked about here, where after Magento 2.2, Grunt does not compile some styles at all due to nested extends.  Perhaps we should open a separate issue?  I'll try to put together a simple example and open a new issue.  Not sure if I'll have time soon, so if somebody else beats me to it, more power to you.

**Comment by giacmir** [comment]
@engcom-backlog-andrii Even if the example reported do not have "impact on page rendering" this does not mean that there is no issue here.

Having two compile systems that produce different ordered rules is a BIG problem in css, where rule position is important (let's remember that the first C of CSS is for "cascading").

In my opinion this issue should be reopened and treated as bug.

**Comment by slackerzz** [comment]
@mverstraete the fact the order of css rules is different between PHP and Grunt is the issue i pointed out, could the issue itself be the solution too?

**Comment by DanielRuf** [comment]
The difference is because of different compilers and different behaviors in the implementations that we use.

Can anybody reproduce this with a simple reproducable setup without Magento but the same dependencies (+ versions)?

**Comment by DanielRuf** [comment]
`oyejorge/less.php` and `grunt-contrib-less`

**Comment by DanielRuf** [comment]
https://github.com/oyejorge/less.php/issues/340

https://github.com/oyejorge/less.php/issues

I propose to switch to another library (for ths PHP stack, probably the official one). We shall not fix a thirdparty dependency which seems to be incomplete.

**Comment by DanielRuf** [comment]
https://github.com/oyejorge/less.php/issues/340

https://github.com/oyejorge/less.php/issues

I propose to switch to another library (for the PHP stack). We shall not fix a thirdparty dependency which seems to be incomplete.

**Comment by DanielRuf** [comment]
http://lessphp.typesettercms.com

**Comment by DanielRuf** [comment]
https://packagist.org/packages/leafo/lessphp
https://github.com/leafo/lessphp
http://leafo.net/lessphp/

**Comment by DanielRuf** [comment]
But it's probably easier to use another pure and single solution instead of having two frontend stacks.

**Comment by DanielRuf** [comment]
https://github.com/gruntjs/grunt-contrib-less/compare/v1.4.1...master

The issue is that no one uses the Grunt version and we just use the PHP version in projects.

Would an upgrade to grunt-contrib-less@2.0.0 fix the issue with the order?

**Comment by jasperzeinstra** [comment]
We're still having this issue on 2.3. The version of `grunt-contrib-less` for 2.3 is set on `^2.0.0`.

Here are two screenshots that show the impact of this issue:
![Screen Shot 2019-03-13 at 16 31 13](https://user-images.githubusercontent.com/9214557/54292176-07437c00-45ae-11e9-8010-8558a4475062.png)
![Screen Shot 2019-03-13 at 16 31 04](https://user-images.githubusercontent.com/9214557/54292177-07437c00-45ae-11e9-8e67-c9dd608d631c.png)

When downgrading `grunt-contrib-less` to version `0.12.0` like @joshdavenport is suggesting the problem is resolved. But it feels very bad to downgrade from `^2.0.0` to `0.12.0`.

This issue has been reported on 28 Oct 2016 and there's still no real solution for this problem. We should be able to do our frontend developing with grunt. Waiting on PHP on every change is taking up way to much time.

@magento-engcom-team could you reply on how you're looking at this issue and what the state is of `MAGETWO-60689` created on 8 november 2016?

**Comment by DanielRuf** [comment]
We can not easily fix this.

Please read all previous commens and also https://github.com/magento/magento2/issues/7231#issuecomment-449398387

**Comment by DanielRuf** [comment]
> We shall not fix a thirdparty dependency which seems to be incomplete.

**Comment by jasperzeinstra** [comment]
From my point of view comment #7231 (comment) is talking about the order of CSS. The issue with nested `&:extend(…)` isn't discussed their. Only @joshdavenport mentioned to lower `grunt-contrib-less`. That solves the issue, but i'm not sure if we're missing any other functionality.

It kind of weird that Magento says use grunt for local frontend development. But when using grunt we're missing some the nested `&:extend(…)` in `&:extend(…)` and we can trust the CSS we're seeing.

**Comment by DrewML** [comment]
Even if we can't fix this immediately, this is a real bug and we should be leaving this ticket open to reflect that.

We as Magento currently offer 2 different ways to compile your LESS:

1. PHP
2. Grunt

Even if these are separate compilers, we should either:

1. Be shipping versions that are compatible with each-other
2. Only ship 1 compiler (either PHP or LessJS/Grunt)

Appreciate the feedback from everyone so far.

**Comment by orlangur** [comment]
@DrewML,

3. Teach developers that result of compilation may be different by nature but strive for elimination of any incompatibilities

Wasn't the plan to switch to SASS by the way?

**Comment by DrewML** [comment]
> Wasn't the plan to switch to SASS by the way?

Hmm, that's news to me. I know people want it, but I haven't seen a concrete proposal for it.

**Comment by orlangur** [comment]
@DrewML there were such discussions much earlier than `architecture` repo was even introduced, here is a community project somebody from Magento should be aware of: https://community.magento.com/t5/Magento-2-x-Less-to-Sass/bd-p/less-to-sass

**Comment by DrewML** [comment]
Ah right - forgot this whole arch process was so new.

To be fairly transparent, I can't imagine us being able to make this change in core in the current platform. Because of back-compat requirements, we'd have to support less _and_ sass for a long period of time, which is difficult since styles are notoriously hard to statically verify.

I will say it's _possible_ we could support SASS running in static content deployment at some future time, but don't think we'll re-write existing themes or publish a new sass theme.

**Comment by DanielRuf** [comment]
> which is difficult since styles are notoriously hard to statically verify

Probably not that hard as we can build ASTs and parse in many different ways.

**Comment by orlangur** [comment]
I may recall it wrong, but if we have a separate theme for PWA the plan was to make it SASS-based.

Why we cannot deprecate LESS core styles in 2.4 and remove in 2.5? LESS compilation then may still be in-place for a longer period.

**Comment by DrewML** [comment]
> Probably not that hard as we can build ASTs and parse in many different ways.

This isn't about validating syntax. The problem would be, any dev doing styling work in core would need to update the luma sass theme and the luma less theme, and then we'd need to always watch for visual regressions across all pages between the two.

> I may recall it wrong, but if we have a separate theme for PWA the plan was to make it SASS-based.

PWA isn't a theme but an entirely new rending and compilation pipeline for the front-end outside of core. It doesn't participate in Magento's CSS compilation at all.

> Why we cannot deprecate LESS core styles in 2.4 and remove in 2.5? LESS compilation then may still be in-place for a longer period.

I'm not sure I see the value in deprecating it. The only way we'd deprecate in core is if we had something new to replace less with, but that would require us maintaining less theme + {insert styling language here} theme for years.

**Comment by orlangur** [comment]
> I'm not sure I see the value in deprecating it

It depends just on opinion "SASS is better than LESS". I have no opinion on this as I don't do HTML/CSS coding.

> but that would require us maintaining less theme + {insert styling language here} theme for years

Why? Just release a SASS theme in 2.4 deprecating LESS theme at the same time. All further fixes are made to SASS only.

**Comment by DrewML** [comment]
> Just release a SASS theme in 2.4 deprecating LESS theme at the same time. All further fixes are made to SASS only.

We would need an _extremely_ good argument for stopping support for a theme almost every m2 store is based upon.

**Comment by MichaelThessel** [comment]
The fact that this is still a problem in 2019 shows that handling dev and production static asset compilation in 2 different ways has been one of the worst design decisions in M2. Unifying this should be a focus of one of the next releases. Working with M2 in it's current state is a massive pain.

**Comment by orlangur** [comment]
@MichaelThessel I would not call it "design decision" as the problem lies in third-party components. M2 tried to cover two main use cases: easily generate assets when you don't even touch frontend VS give you an ability to use original JS library. And did it's job pretty well - just that you need to understand natural limitations of both approaches.

The problem is even deeper and lies in LESS itself which, unlike SASS, does not even have a specification.

I don't think if `less.php` will be dropped backend developers will become really happy having to setup additional tooling. What would you propose as a "unifying"?

My understanding is that "server side compilation" based on `lessphp` is not production ready in it's current state (and should not be - corresponding library is kinda abandoned) thus it should only be available for development purposes with a big WARNING to use `lessjs` when you work with frontend intensively.

**Comment by hostep** [comment]
> My understanding is that "server side compilation" based on lessphp is not production ready in it's current state (and should not be - corresponding library is kinda abandoned)

Actually the `less.php` library was taken over a while ago [by wikimedia](https://github.com/oyejorge/less.php/issues/329#issuecomment-430122809): https://github.com/wikimedia/less.php/. Recently support for PHP 7.3 & 7.4 was added for example. So it's still kind of maintained. If there are issues with the library, PR's to it will most likely get reviewed and accepted.

Magento 2.3.3 will ship with this new fork btw: https://github.com/magento/magento2/commit/9a4332cb18945772c1458d53cd1a9c191c3f7a10

**Comment by MichaelThessel** [comment]
@orlangur Since the introduction of CSS pre-processors Magento is the only project that I have come across that has 2 different ways to compile CSS. Any other project uses the same compilation approach for production and development. The regular way is to just compile it the same way you do on your dev system during CI and then bundle the compiled CSS with the rest of the build and deploy it to production. The fact that I have to shut down my site for 2 minutes to compile my static assets would be considered **insane** in any other project. Just look at this thread and you will see that its not working for tons of people.

**Comment by orlangur** [comment]
@hostep thanks for sharing! https://www.google.com/search?q=lessphp only showed me `oyejorge` and `leafo`, both unmaintained.

@MichaelThessel, sure, I see. Materializing compiled files somewhere is really robust, could be used even for generated PHP code.

**Comment by DrewML** [comment]
Thanks for the good discussion, y'all.

Just to be incredibly clear from my side on behalf of Magento, I consider this to be a bug. Styles (without a code change) should not change between dev and prod.

We should be using 1 compiler, period. Whether it's the PHP or the JS version is slightly less important.

For the short-term, until we can address this at a code level, we need to make it super clear that you should be using the same compiler for both dev and prod.

Long term, the preference would be for us to move to the official compiler. The majority of front-end tools in 2019 are authored in JS, so attempting to keep things in PHP for the sake of consistency is a losing battle.

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @engcom-Charlie 
 Thank you for verifying the issue. Based on the provided information internal tickets `MC-22208` were created

**Issue Available**: @engcom-Charlie, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by mrtuvn** [comment]
The difference obviously occurs
Here is one example

When use server side compile  bear in minds rules/mixins in file _extends.less will not generate to output css. It's work as we expected. Default magento import all the them as (reference)
https://github.com/magento/magento2/blob/2.4-develop/app/design/frontend/Magento/blank/web/css/styles-l.less#L17

http://lesscss.org/features/#import-atrules-feature-reference

But if we use npm less compile all rules in _extends still included in output

**Comment by Aquive** [comment]
Is this fixed? I think dev compilation should be the same result as what is produced in production mode no matter which tool. We shouldn't be saying things like "it works on my machine". Work on dev with grunt, deploy to production and just see same result...

**Comment by engcom-November** [comment]
As this is quite a old issue, we verified this issue on latest 2.4-develop branch and the issue is not reproducible, it's been fixed.

Made the css generation to `server side compilation` and enabled developer mode.

There were no difference in grunt and server side compilation:
**styles-m.css**
<img width="1286" alt="image" src="https://github.com/user-attachments/assets/74af1921-87f7-4be4-bfde-243f5c23428d">

**styles-l.css**
<img width="1279" alt="image" src="https://github.com/user-attachments/assets/db1d5f97-aadf-490f-9cb7-86396ab886c6">

Thank you.

**Comment by hostep** [comment]
@engcom-November: I'm sure you executed your testing incorrectly.

I do still see differences, but the situation has improved a lot since the issue got reported, mostly due to the `wikimedia/less.php` composer package to be at version 5 (not released yet, is planned for Magento 2.4.8), which is compatible with less.js v3. The difference can be explained because grunt uses less.js v4 since Magento 2.4.7, so differences are still to be expected.
If we want to eliminate differences, we either need to wait until `wikimedia/less.php` comes out with another update to get it in line with less.js v4, or we need to downgrade less.js v4 back to v3 for grunt compilation.
But not sure if that is all needed, I would only do this if somebody comes up with a true example of what has a completely different result on the frontend between grunt & the less.php library.

Anyway, proper steps to reproduce:

1. Have a fresh clean and working Magento 2.4-develop instance
2. `cp package.json.sample package.json`
3. `cp Gruntfile.js.sample Gruntfile.js`
4. `npm install`
5. `bin/magento setup:upgrade` (to clear all generated files)
6. `grunt clean:luma && grunt exec:luma && grunt less:luma`
7. `cp -r pub/static /tmp/magento-static-grunt`
8. `bin/magento setup:upgrade` (to clear all generated files)
9. `bin/magento setup:static-content:deploy -f --no-parent --theme Magento/luma en_US`
10. `diff -u /tmp/magento-static-grunt/frontend/Magento/luma/en_US/css/styles-l.css pub/static/frontend/Magento/luma/en_US/css/styles-l.css`
11. `diff -u /tmp/magento-static-grunt/frontend/Magento/luma/en_US/css/styles-m.css pub/static/frontend/Magento/luma/en_US/css/styles-m.css`

The diff in `styles-l.css` is very minor:
```diff
--- /tmp/magento-static-grunt/frontend/Magento/luma/en_US/css/styles-l.css	2024-10-23 22:40:09
+++ pub/static/frontend/Magento/luma/en_US/css/styles-l.css	2024-10-23 22:41:09
@@ -2112,7 +2112,7 @@
     font-family: 'Open Sans', 'Helvetica Neue', Helvetica, Arial, sans-serif;
     font-size: 14px;
     height: 32px;
-    line-height: 1.428571429;
+    line-height: 1.42857143;
     padding: 0 9px;
     vertical-align: baseline;
     width: 100%;
\ No newline at end of file
@@ -3175,4 +3175,3 @@
     margin-left: 85px;
   }
 }
-/*# sourceMappingURL=styles-l.css.map */
\ No newline at end of file
```

The one in `styles-m.css` is much bigger, so I'm no going to post it here.
